clustering/ejb/WFLY-7628_Distributed_EJB_Timers.adoc

```diff
@@ -56,87 +59,108 @@ WildFly needs an persistent timer service implementation suitable for use in a c
+* Implement an distributed implementation of the `TimerService` SPI.
+** Add a `org.wildfly.clustering.ejb.timer` SPI to the `wildfly-clustering-ejb-spi` module to facilitate multiple distributed timer manager implementations.
+** Create `DistributedTimerService`, a distributed `TimerService` implementation to the ejb3 subsystem that delegates to the `wildfly-clustering-ejb-spi` for timer management and scheduling.
+** Add an Infinispan implementation of the `org.wildfly.clustering.ejb.timer` SPI that assigns execution ownership of timers to cluster members based on consistent hashing of the timer identifier.
+*** Leverage the existing `LocalScheduler` implementation within `wildfly-clustering-ee-cache` for timer scheduling
+*** Leverage the existing `PrimaryOwnerScheduler` decorator that uses consistent hashing of the timer identifier to distribute timer execution among cluster members
+* Add a management resource and capability to the distributable-ejb subsystem to configure the Infinispan timer management implementation.
+* Make the `default-data-store` attribute of the timer service resource optional.
+** Require the default distributable timer management capability when undefined.
+** Modify the `TimerService` DUP to use a `CompositeTimerService` in the event that the server is configured to use persistent timer management, that delegates to the existing `TimerServiceImpl` for use with transient timers and the `DistributableTimerService` for use with persistent timers.
 
 == Security Considerations
 
-////
-Identification if any security implications that may need to be considered with this feature
-or a confirmation that there are no security implications to consider.
-////
+N/A
 
 == Test Plan
 
-* All existing TimerService tests must pass against the new persistent timer implementation, including the EJB TCK.
-* Existing distributed EJB timer tests that use the JDBC timer-store within an EAP cluster may need to be modified since timers may execute on different cluster members than expected by the tests.
-* New tests will be added to the clustering integration testsuite to validate correctness of behavior with a cluster, e.g. failover behavior, etc.
+* All existing `TimerService` tests must pass against the new persistent timer implementation, including the Jakarta Enterprise Beans TCK.
+* New tests will be added to the clustering integration testsuite to validate correctness of behavior of persistent timers with a cluster, including:
+** Validate that auto-timers are only ever created on a single cluster member.
+** Validate normal expiration of auto-timers and manually created single action, interval, and calendar persistent timers within stable cluster.
+** Validate that all timer types continue to trigger expirations without duplication after the cluster member that created them is stopped.
+** Validate that all timer types failback to original cluster member without duplication following restart.
```

Comment:
The last two bullet should cover failover scenarios, correct?

Reply:
BTW - one more question on this. How will be the configuration tested? i.e. attributes and operations on new resources?

Reply:
> The last two bullet should cover failover scenarios, correct?

Yes.

> How will be the configuration tested? i.e. attributes and operations on new resources?

These are covered by the distributable-ejb subsystem tests.

Reply:
> These are covered by the distributable-ejb subsystem tests.

the following one as well?

`/subsystem=ejb3/service=timer-service:undefine-attribute(name=default-data-store)`

It is not about the distributable-ejb subsystem, hence my question. Just double checking.

Reply:
Yes.